sources/scraping/queue.py

```diff
@@ -0,0 +1,81 @@
+import typing as t
+
+from queue import Empty, Queue
+
+from dlt.common import logger
+
+
+# Please read more at https://mypy.readthedocs.io/en/stable/runtime_troubles.html#not-generic-runtime
+T = t.TypeVar("T")
+
+if t.TYPE_CHECKING:
+
+    class _Queue(Queue[T]):
+        pass
+
+else:
+
+    class _Queue(Queue, t.Generic[T]):
+        pass
+
+
+class QueueClosedError(Exception):
+    pass
+
+
+class ScrapingQueue(_Queue):
+    def __init__(
+        self,
+        maxsize: int = 0,
+        batch_size: int = 10,
+        read_timeout: float = 1.0,
+    ) -> None:
+        super().__init__(maxsize)
+        self.batch_size = batch_size
+        self.read_timeout = read_timeout
+        self._is_closed = False
+
+    def get(self, block: bool = True, timeout: t.Optional[float] = None) -> T:
+        timeout = timeout if timeout else self.read_timeout
+        batch: t.List[T] = []
+        while True:
+            if len(batch) >= self.batch_size:
+                return batch
+
+            try:
+                if self.is_closed:
+                    raise QueueClosedError("Queue is closed")
+
+                item = super().get(block, timeout=timeout)
+                batch.append(item)
+
+                # Mark task as completed
+                self.task_done()
+            except Empty:
+                logger.info(f"Queue has been empty for {timeout}s...")
+                return batch
+            except QueueClosedError:
```

Comment:
@rudolfix should we catch both exceptions at once, I mean to merge `except Empty, QueueClosedError`?